core/src/main/java/com/graphhopper/GraphHopper.java

```diff
@@ -702,16 +713,20 @@ public boolean load(String graphHopperFolder) {
 
         setGraphHopperLocation(graphHopperFolder);
 
-        if (encodingManager == null)
-            setEncodingManager(EncodingManager.create(encodedValueFactory, flagEncoderFactory, ghLocation));
+        if (encodingManager == null) {
+            StorableProperties properties = EncodingManager.loadProperties(ghLocation);
+            encodingManager = properties == null ?
+                    createEncodingManager()
+                    : EncodingManager.create(encodedValueFactory, flagEncoderFactory, properties);
```

Comment:
Before a missing EncodingManager meant that it should load it from an existing graph folder. 
Now a missing EncodingManager also means that the EncodingManager should be implicitly created from the specified profiles.

Reply:
This only creates the EncodingManager from the existing graph if it hasn't been set earlier. This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case

I suggest that we always create an EntityManager from the graph if present and throw an exception if their is a mismatch.

Reply:
@karussell  Maybe this code becomes more clear if we move the check for `graph.encoded_values` and `graph.flag_encoders` properties from loadProperties to here?

@otbutz I thought the point is that we cannot set the EncodingManager anymore?

Reply:
The EncodingManager can still be populated via `init(GraphHopperConfig)` which is used to create it from the GraphHopperConfig:

https://github.com/graphhopper/graphhopper/blob/785dac682da211411f648849d3d611ac1f3a96f8/core/src/main/java/com/graphhopper/GraphHopper.java#L457-L460

Reply:
Ah ok I see. Yes this way we can set the encoding manager.

>  This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case

What exactly is the problem here? Aren't we supposed to use an EncodingManager that is compatible with the stored information?

Reply:
> This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case

I do not see how #2243 would be affected from this PR. Can you explain. To my knowledge all use cases are still possible as before.

Reply:
See https://github.com/graphhopper/graphhopper/pull/2243#issuecomment-792769503

tl;dr: the solution proposed in #2243 is currently not working because the EntityManager isn't created from the graph if it was set before by `init()`.

> To my knowledge all use cases are still possible as before.

Your PR doesn't change this behavior but i think it makes sense to change it if we restructure EM handling.

Reply:
> Ah ok I see. Yes this way we can set the encoding manager.
> 
> > This is problematic for things like #2243 where we might have stored additional information in the graph which are not present in the EntityManager being passed in. e.g the encountered String values in this case
> 
> What exactly is the problem here? Aren't we supposed to use an EncodingManager that is compatible with the stored information?

The current implementation of the StringEncodedValue only stores **encountered** values in order to safe space. The problem with this approach is that we need to ensure that the EncodingManager is created from the graph in order to retrieve the list of encountered values.

For compatibility checks it's enough to compare the usual EV properties like number of bits etc.

Reply:
Yes it would be better to load the EncodingManager from the graph. Unfortunately we do not do this yet and this is probably a bigger endeavour. The intention of this PR is that there will be an `is_subnetwork` encoded value that we need per profile (not per encoder). Since there is already `GraphHopper#setProfiles` it would be redundant to also keep `GraphHopper#setEncodingManager` so the idea was to remove `setEncodingManager` and create the encoding manager based on the given profiles. But this certainly won't be the end of the story and we will do more refactoring of the GraphHopper class.

Reply:
> Unfortunately we do not do this yet and this is probably a bigger endeavour. 

Is it? Wouldn't the following be enough to make it work:

```java
StorableProperties properties = EncodingManager.loadProperties(ghLocation);
if (properties != null) {
     EncodingManager graphEncodingManager = EncodingManager.create(encodedValueFactory, flagEncoderFactory, properties);
     if (encodingManager != null && graphEncodingManager.equals(encodingManager)) {
          throw ...
     }
     encodingManager = graphEncodingManager;
} else if (encodingManager == null) {
    encodingManager = createEncodingManager();
}
```

Reply:
No, I don't think so. For example we do not store information about profiles.